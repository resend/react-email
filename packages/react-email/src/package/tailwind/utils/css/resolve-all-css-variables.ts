import type { Node, Root } from 'patched-postcss';
import type AtRule from 'patched-postcss/lib/at-rule';
import Declaration from 'patched-postcss/lib/declaration';
import Rule from 'patched-postcss/lib/rule';
import { removeIfEmptyRecursively } from './remove-if-empty-recursively';

const isRule = (node: Node | undefined): node is Rule => {
  return node?.type === 'rule';
};

const isAtRule = (node: Node | undefined): node is AtRule => {
  return node?.type === 'atrule';
};

const doesNodeShareVariableWith = (
  node: Node | undefined,
  nodeToShareWith: Node | undefined,
) => {
  if (isRule(node) && isRule(nodeToShareWith)) {
    return (
      node.selector === nodeToShareWith.selector ||
      node.selector.includes('*') ||
      node.selector.includes(':root')
    );
  }

  return node === nodeToShareWith;
};

const findVariableDeclarations = (root: Root, rule: Rule, variable: string) => {
  const variableDeclarations = new Set<Declaration>();
  root.walkDecls((declaration) => {
    if (/--[^\s]+/.test(declaration.prop)) {
      if (
        `var(${declaration.prop})` === variable &&
        doesNodeShareVariableWith(declaration.parent, rule)
      ) {
        variableDeclarations.add(declaration);
      }
    }
  });
  return variableDeclarations;
};

export const resolveAllCSSVariables = (root: Root) => {
  root.walkRules((rule) => {
    const declarationsForAtRules = new Map<AtRule, Set<Declaration>>();
    const valueReplacingInformation = new Set<{
      declaration: Declaration;
      newValue: string;
    }>();

    rule.walkDecls((declaration) => {
      if (/var\(--[^\s)]+\)/.test(declaration.value)) {
        /**
         * @example ['var(--width)', 'var(--length)']
         */
        const variablesUsed = [
          ...(/var\(--[^\s)]+\)/gm.exec(declaration.value) ?? []),
        ];
        for (const variableUsed of variablesUsed) {
          const variableDeclarations = findVariableDeclarations(
            root,
            rule,
            variableUsed,
          );

          for (const variableDeclaration of variableDeclarations) {
            if (
              variableDeclaration.parent &&
              isAtRule(variableDeclaration.parent.parent) &&
              variableDeclaration.parent !== rule
            ) {
              const atRule = variableDeclaration.parent.parent;

              const declarationWithoutVariable = new Declaration({
                important: declaration.important,
                prop: declaration.prop,
                value: declaration.value.replaceAll(
                  variableUsed,
                  variableDeclaration.value,
                ),
              });

              const declarationForAtRule = declarationsForAtRules.get(atRule);
              if (declarationForAtRule) {
                declarationForAtRule.add(declarationWithoutVariable);
              } else {
                declarationsForAtRules.set(
                  atRule,
                  new Set([declarationWithoutVariable]),
                );
              }
              break;
            }

            valueReplacingInformation.add({
              declaration,
              newValue: declaration.value.replaceAll(
                variableUsed,
                variableDeclaration.value,
              ),
            });
          }
        }
      }
    });

    for (const { declaration, newValue } of valueReplacingInformation) {
      declaration.value = newValue;
    }

    for (const [atRule, declarations] of declarationsForAtRules.entries()) {
      const equivalentRule = new Rule({
        selector: rule.selector,
        selectors: rule.selectors,
        nodes: [...declarations],
      });
      atRule.append(equivalentRule);
    }
  });

  root.walkDecls((decl) => {
    if (/--[^\s]+/.test(decl.prop)) {
      const parent = decl.parent;
      decl.remove();
      if (parent) {
        removeIfEmptyRecursively(parent);
      }
    }
  });

  return root;
};
