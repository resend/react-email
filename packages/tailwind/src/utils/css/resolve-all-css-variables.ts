import {
  AtRule,
  type Declaration,
  type Root,
  Rule,
} from 'postcss';
import { removeIfEmptyRecursively } from './remove-if-empty-recursively';

interface VariableUse {
  declaration: Declaration;
  variable: string;
}

interface VariableDefinition {
  declaration: Declaration;
  variable: string;
  definition: string;
}

const doSelectorsIntersect = (first: string, second: string): boolean => {
  if (first === second) {
    return true;
  }

  if (first.includes(':root') || second.includes(':root')) {
    return true;
  }

  if (first.includes('*') || second.includes('*')) {
    return true;
  }

  return false;
};

export const resolveAllCSSVariables = (root: Root) => {
  const variableDefinitions = new Set<VariableDefinition>();
  const variableUses = new Set<VariableUse>();

  root.walkDecls((declaration) => {
    if (/--[^\s]+/.test(declaration.prop)) {
      variableDefinitions.add({
        declaration,
        variable: `var(${declaration.prop})`,
        definition: declaration.value,
      });
    } else {
      const variablesUsed = [
        ...declaration.value.matchAll(/var\(--[^\s)]+\)/gm),
      ].map((match) => match.toString());

      for (const variable of variablesUsed) {
        variableUses.add({
          declaration,
          variable,
        });
      }
    }
  });

  for (const definition of variableDefinitions) {
    for (const use of variableUses) {
      if (use.variable !== definition.variable) {
        continue;
      }

      if (
        use.declaration.parent instanceof AtRule &&
        use.declaration.parent.parent instanceof Rule &&
        definition.declaration.parent instanceof Rule &&
        doSelectorsIntersect(
          use.declaration.parent.parent.selector,
          definition.declaration.parent.selector,
        )
      ) {
        use.declaration.value = use.declaration.value.replaceAll(
          use.variable,
          definition.definition,
        );
        continue;
      }

      if (
        use.declaration.parent instanceof Rule &&
        definition.declaration.parent instanceof Rule &&
        doSelectorsIntersect(
          use.declaration.parent.selector,
          definition.declaration.parent.selector,
        )
      ) {
        use.declaration.value = use.declaration.value.replaceAll(
          use.variable,
          definition.definition,
        );
      }
    }
  }

  // root.walkRules((rule) => {
  //   const declarationsForMediaQueries = new Map<AtRule, Set<Declaration>>();
  //   const valueReplacingInformation = new Set<{
  //     declaration: Declaration;
  //     replacing: string;
  //     replacement: string;
  //   }>();
  //
  //   rule.walkDecls((declaration) => {
  //     if () {
  //       /**
  //        * @example ['var(--width)', 'var(--length)']
  //        */
  //       const variablesUsed = [
  //         ...declaration.value.matchAll(/var\(--[^\s)]+\)/gm),
  //       ].map((match) => match.toString());
  //
  //       root.walkDecls((otherDecl) => {
  //         if () {
  //           const variable = `var(${otherDecl.prop})`;
  //           if (
  //             variablesUsed?.includes(variable) &&
  //             areSelectingTheSame(declaration.parent, otherDecl.parent)
  //           ) {
  //             if (
  //               otherDecl.parent?.parent instanceof AtRule &&
  //               otherDecl.parent?.parent.name === 'media' &&
  //               otherDecl.parent !== declaration.parent
  //             ) {
  //               const atRule = otherDecl.parent.parent;
  //
  //               const clonedDeclaration = createDeclaration();
  //               clonedDeclaration.prop = declaration.prop;
  //               clonedDeclaration.value = declaration.value.replaceAll(
  //                 variable,
  //                 otherDecl.value,
  //               );
  //               clonedDeclaration.important = declaration.important;
  //
  //               const declarationForAtRule =
  //                 declarationsForMediaQueries.get(atRule);
  //               if (declarationForAtRule) {
  //                 declarationForAtRule.add(clonedDeclaration);
  //               } else {
  //                 declarationsForMediaQueries.set(
  //                   otherDecl.parent.parent,
  //                   new Set([clonedDeclaration]),
  //                 );
  //               }
  //               return;
  //             }
  //
  //             valueReplacingInformation.add({
  //               declaration,
  //               replacing: variable,
  //               replacement: otherDecl.value,
  //             });
  //           }
  //         }
  //       });
  //     }
  //   });
  //
  //   for (const {
  //     declaration,
  //     replacing,
  //     replacement,
  //   } of valueReplacingInformation) {
  //     declaration.value = declaration.value.replaceAll(replacing, replacement);
  //   }
  //
  //   for (const [
  //     atRule,
  //     declarations,
  //   ] of declarationsForMediaQueries.entries()) {
  //     const equivalentRule = createRule();
  //     equivalentRule.selector = rule.selector;
  //     equivalentRule.append(...declarations);
  //
  //     atRule.append(equivalentRule);
  //   }
  // });

  for (const definition of variableDefinitions) {
    const parent = definition.declaration.parent;
    definition.declaration.remove();
    if (parent) {
      removeIfEmptyRecursively(parent);
    }
  }

  return root;
};
