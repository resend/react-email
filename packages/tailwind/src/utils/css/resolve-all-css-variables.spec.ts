import { parse } from 'postcss';
import { resolveAllCSSVariables } from './resolve-all-css-variables';

describe('resolveAllCSSVariables', () => {
  it('should work with simple css variables on a :root', () => {
    const root = parse(`:root {
  --width: 100px;
}

.box {
  width: var(--width);
}`);

    expect(resolveAllCSSVariables(root).toString()).toMatchSnapshot();
  });

  it('should work for variables across different CSS layers', () => {
    const root = parse(`@layer base {
      :root {
        --width: 100px;
      }
    }

    @layer utilities {
      .box {
        width: var(--width);
      }
    }`);

    expect(resolveAllCSSVariables(root).toString()).toMatchSnapshot();
  });

  it('should work with multiple variables in the same declaration', () => {
    const root = parse(`:root {
      --top: 101px;
      --bottom: 102px;
      --right: 103px;
      --left: 104px;
    }

    .box {
      margin: var(--top) var(--right) var(--bottom) var(--left);
    }`);

    expect(resolveAllCSSVariables(root).toString()).toMatchSnapshot();
  });

  it('should keep variable usages if it cant find their declaration', () => {
    const root = parse(`.box {
  width: var(--width);
}`);

    expect(resolveAllCSSVariables(root).toString()).toMatchSnapshot();
  });

  it('should work with variables set in the same rule', () => {
    const root = parse(`.box {
  --width: 200px;
  width: var(--width);
}

@media (min-width: 1280px) {
  .xl\\:bg-green-500 {
    --tw-bg-opacity: 1;
    background-color: rgb(34 197 94 / var(--tw-bg-opacity))
  }
}
`);
    expect(resolveAllCSSVariables(root).toString()).toMatchSnapshot();
  });

  it('should work with a variable set in a layer, and used in another through a media query', () => {
    const root = parse(`@layer theme {
  :root {
    --color-blue-300: blue;
  }
}

@layer utilities {
  .sm\:bg-blue-300 {
    @media (width >= 40rem) {
      background-color: var(--color-blue-300);
    }
  }
}`);
    expect(resolveAllCSSVariables(root).toString()).toMatchSnapshot();
  });

  // this behavior is not supported anymore, since it doesn't seem like tailwindcss actually generates any CSS that uses the pattern of defining css variables from inside media queries
  //   it.only('should work with different values between media queries', () => {
  //     const root = parse(`:root {
  //   --width: 100px;
  // }
  //
  // @media (max-width: 1000px) {
  //   :root {
  //     --width: 200px;
  //   }
  // }
  //
  // .box {
  //   width: var(--width);
  // }`);
  //     expect(
  //       resolveAllCSSVariables(root).toString(),
  //     ).toBe(`@media (max-width: 1000px) {
  //   .box {
  //     width: 200px;
  //   }
  // }
  //
  // .box {
  //   width: 100px;
  // }`);
  //   });
});
