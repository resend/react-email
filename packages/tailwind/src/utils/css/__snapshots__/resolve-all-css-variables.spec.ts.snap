// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`resolveAllCSSVariables > should keep variable usages if it cant find their declaration 1`] = `
".box {
  width: var(--width);
}"
`;

exports[`resolveAllCSSVariables > should work for variables across different CSS layers 1`] = `
"@layer utilities {
      .box {
        width: 100px;
      }
    }"
`;

exports[`resolveAllCSSVariables > should work with a variable set in a layer, and used in another through a media query 1`] = `
"@layer utilities {
  .sm:bg-blue-300 {
    @media (width >= 40rem) {
      background-color: blue;
    }
  }
}"
`;

exports[`resolveAllCSSVariables > should work with multiple variables in the same declaration 1`] = `
".box {
      margin: 101px 103px 102px 104px;
    }"
`;

exports[`resolveAllCSSVariables > should work with simple css variables on a :root 1`] = `
".box {
  width: 100px;
}"
`;

exports[`resolveAllCSSVariables > should work with variables set in the same rule 1`] = `
".box {
  width: 200px;
}

@media (min-width: 1280px) {
  .xl\\:bg-green-500 {
    background-color: rgb(34 197 94 / 1)
  }
}
"
`;
